numpy/lib/_index_tricks_impl.py

```diff
@@ -641,7 +643,406 @@ def __iter__(self):
+    """
+    An N-dimensional range-like object for indexing arrays.
+
+    The ndrange type represents an immutable sequence of tuples of numbers and
+    can be used for looping over specific multi-dimensional indicies in
+    ``for`` loops. Can be called as ``ndrange(stop)`` or
+    ``ndrange(start, stop, [step])``, see parameters below.
+
+    When iterated, ndrange produces a sequence of tuples of integers
+    `(i_0, i_1, ..., i_Nminus1)` as produced by nested for-loops over ranges.
+    For instance, for dimension 3 this sequence is identical to that produced
+    by:
+
+        ((i_0, i_1, i_2) for i_0 in range(start[0], stop[0], step[0])
+                         for i_1 in range(start[1], stop[1], step[1])
+                         for i_2 in range(start[2], stop[2], step[2]))
+
+    While the ndrange range iterates as a flat sequence, it can be sliced along
+    multiple axes. The ndrange object also has attributes ``ndim`` and
+    ``shape`` corresponding to the number of dimensions and  the number of
+    indexed elements along each dimension respectively.
+
+    The ndrange object is not designed as a substitute for vectorized numeric
+    code. Whenever possible, vectorized routines should be used for
+    mathematical operations. ndrange is provided primarily for convenient for
+    iteration over multi-dimensional collections when no vectorized primitives
+    exist.
+
+    Like the Python 3 range object, ndrange objects implement the
+    ``collections.abc.Sequence`` ABC and provide features such as membership
+    tests, element index lookup, slicing, and support for negative indices.
+    See the `python class range documentation
+    <https://docs.python.org/3/library/stdtypes.html#range>`_ for more details.
+
+    Testing ndrange objects for equality with ``==`` and ``!=`` compares them
+    as sequences. See the examples for more details.
+
+    If ndrange is called with a single positional argument, that argument
+    is assumed to be ``stop`` and ``start`` defaults to a tuple of zeros.
+
+    Parameters
+    ----------
+    stop: tuple of ints, or a single int
+        Specifies the stop parameter along each axis.
+    start: tuple of ints, or a single int, optional
+        If provided, must be the same length as stop. Specifies the start
+        parameter along each axis.
+    step: tuple of ints, or a single int, optional
+        If provided, must be the same length as stop. Specifies the increment
+        along each axis. If step is not provided, it defaults to a tuple of
+        ones.
+
+    When a single position parameter is provided, ndrange assumes it is
+    the ``stop`` parameter and the ndrange constructor behaves as though it has
+    the signature ``ndrange(stop)``. In all other cases, ``ndrange``
+    behaves as though the signalture is ``ndrange(start, stop[, step])``.
+
+    See Also
+    --------
+    flatiter, ndenumerate, ndindex
+
+    Notes
+    -----
+    .. versionadded:: 2.2.0
```

Comment:
is this correct??

Reply:
Sure, assuming we review and merge before the 2.2 release :)